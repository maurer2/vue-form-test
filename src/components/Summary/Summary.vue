<template>
  <div class="summary">
    <h2 class="title">
      Summary
    </h2>
    <p class="message" v-if="entries.length === 0">
      No entries found.
    </p>
    <ul class="list" v-else>
      <template v-for="(entry, index) in entries">
        <li class="entry" :class="{ 'entry--is-active' : (index === activeEntry) }" :key="index">
          <button
            type="button"
            class="button"
            :class="{ 'button--is-active' : (index === activeEntry) }"
            @click="handleClick(index)"
          >
            {{ entry.firstName }} {{ entry.lastName }}
          </button>

          <template v-if="activeEntry === index">
            <dl class="detail">
              <dd class="key">
                First name
              </dd>
              <dd class="value">
                {{ entry.firstName }}
              </dd>
            </dl>

            <dl class="detail">
              <dd class="key">
                Last name
              </dd>
              <dd class="value">
                {{ entry.lastName }}
              </dd>
            </dl>

            <dl class="detail">
              <dd class="key">
                E-mail
              </dd>
              <dd class="value">
                {{ entry.email }}
              </dd>
            </dl>

            <dl class="detail" v-if="entry.customerQuery.length > 0">
              <dd class="key">
                Query
              </dd>
              <dd class="value">
                {{ entry.customerQuery }}
              </dd>
            </dl>
          </template>
        </li>
      </template>
    </ul>
  </div>
</template>

<script lang="ts">
import { Vue, Component } from 'vue-property-decorator';

@Component
export default class Field extends Vue {
  private activeEntry: number = (this.entries.length > 0) ? 0 : -1;

  get entries() {
    return this.$store.state.entries;
  }

  handleClick(index: number): void {
    if (index === this.activeEntry) {
      this.activeEntry = -1;
      return;
    }

    this.activeEntry = index;
  }
}
</script>

<style scoped lang="scss">
.title {
  margin-top: 0;
  margin-bottom: 0.75rem;
  padding-bottom: 0.75rem;
  border-bottom: 1px dashed $blue;
}

.list {
  display: flex;
  margin: 1rem 0;
  margin-top: -0.75rem;
  margin-left: -0.75rem;
  padding: 0;
  flex-wrap: wrap;
  list-style: none;
}

.entry {
  display: flex;
  margin-top: 0.75rem;
  margin-left: 0.75rem;
  flex-direction: column;

  &--is-active {
    flex-grow: 1;
    flex-basis: 100%;
    flex-shrink: 0;
  }
}

.button {
  position: relative;
  padding-right: 2.5rem;

  &:after {
    content: '';
    position: absolute;
    right: 1rem;
    top: 50%;
    width: 0;
    height: 0;
    border-left: 0.5rem solid transparent;
    border-right: 0.5rem solid transparent;
    border-top: 0.5rem solid $white;
    transform: translateY(-50%);
  }

  &--is-active:after {
    border-top: 0;
    border-bottom: 0.5rem solid $white;
  }
}

.detail {
  display: flex;
  margin: 0.5rem 1rem 0 1rem;
}

.key {
  margin: 0;
  flex-basis: 6rem;
  font-weight: bold;
}

.value {
  margin: 0rem;
}
</style>
