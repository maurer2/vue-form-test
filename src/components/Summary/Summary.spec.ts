import { shallowMount, createLocalVue } from '@vue/test-utils';
import Vuex from 'vuex';
import Summary from './Summary.vue';

const localVue = createLocalVue();
const mockedClickEvent = jest.fn();

localVue.use(Vuex);

describe('Summary', () => {
  let wrapper: any;
  let store: any;

  beforeEach(() => {
    store = new Vuex.Store({
      state: {
        entries: [
          {
            firstName: 'Cat',
            lastName: 'Damon',
            email: 'catD@mon.com',
            customerQuery: '',
          },
          {
            firstName: 'Catrick',
            lastName: 'Swayze',
            email: 'catrick@swayze.com',
            customerQuery: '',
          },
        ],
      },
    });

    wrapper = shallowMount(Summary, {
      propsData: {},
      store,
      localVue,
    });
  });

  afterEach(() => {
    wrapper.destroy();
  });

  test('Component matches snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  test('Summary contains child elements', () => {
    expect(wrapper.find('.summary').exists()).toBe(true);
    expect(wrapper.find('.title').exists()).toBe(true);
    expect(wrapper.find('.list').exists()).toBe(true);
  });

  test('Summary element with the same index as activeEntry has .entry--is-active class', () => {
    wrapper.setData({ activeEntry: 1 });
    expect(wrapper.findAll('.entry').at(1).find('.entry--is-active').exists()).toBe(true);
    expect(wrapper.findAll('.entry').at(0).find('.entry--is-active').exists()).not.toBe(true);

    wrapper.setData({ activeEntry: 0 });
    expect(wrapper.findAll('.entry').at(0).find('.entry--is-active').exists()).toBe(true);
    expect(wrapper.findAll('.entry').at(1).find('.entry--is-active').exists()).not.toBe(true);
  });

  test('Summary has only one active element', () => {
    expect(wrapper.findAll('.entry--is-active').length).toBe(1);
  });

  test('Summary handleClick changes class to inactive when button is active', () => {
    wrapper.setData({ activeEntry: 1 });
    wrapper.findAll('.entry').at(1).find('.button').trigger('click');

    expect(wrapper.vm.activeEntry).toBe(-1);
  });

  test('Summary handleClick changes class to active when button is inactive', () => {
    wrapper.setData({ activeEntry: 0 });
    wrapper.findAll('.entry').at(1).find('.button').trigger('click');

    expect(wrapper.vm.activeEntry).toBe(1);
  });

  test('Summary handleClick method is called on click', () => {
    wrapper.setMethods({ handleClick: mockedClickEvent });
    wrapper.find('.entry > .button').trigger('click');

    expect(mockedClickEvent).toBeCalled();
  });
});
