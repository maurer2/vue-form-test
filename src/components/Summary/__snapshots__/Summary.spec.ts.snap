// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Summary Component matches snapshot 1`] = `
<div class="summary">
  <h2 class="title">
    Summary
  </h2>
  <ul class="list">
    <li class="entry entry--is-active"><button type="button" class="button button--is-active">
        Cat Damon
      </button>
      <dl class="detail">
        <dd class="key">
          First name
        </dd>
        <dd class="value">
          Cat
        </dd>
      </dl>
      <dl class="detail">
        <dd class="key">
          Last name
        </dd>
        <dd class="value">
          Damon
        </dd>
      </dl>
      <dl class="detail">
        <dd class="key">
          E-mail
        </dd>
        <dd class="value">
          catD@mon.com
        </dd>
      </dl>
      <!---->
    </li>
    <li class="entry"><button type="button" class="button">
        Catrick Swayze
      </button>
      <!---->
    </li>
  </ul>
</div>
`;
